woocommerce/woocommerce-android
# [Better Errors] Introduce Connectivity Tool logic

Summary
==========
Fix issue #10799 by introducing the Connectivity Tool entities, implementation and behavior.

There are four connectivity checks:

- Internet connection check. Simply check internet availability.
- WordPress.com Servers check. Will ping the `what's new` endpoint hosted in the WPCOM servers to verify the overall availability.
- Store connection check. Will ping the Store System Report (SSR) endpoint to make sure that the store is up and running.
- Store Orders check. Will ping the Store orders count endpoint to verify there are response from the Orders features.

Please notice that the actual UI for this view will be introduced in https://github.com/woocommerce/woocommerce-android/pull/11025, the one introduced here is barebones and only serves the purpose of testing the actual Connectivity checks.

⚠️ For a better review navigation and size, All Connectivity use cases unit tests are in a separate PR: https://github.com/woocommerce/woocommerce-android/pull/11024

How to Test
==========
### Code changes
Since the timeout error is hard to reproduce, and this test scenario requires two timeout events one after the other, I suggest applying this patch for better testing:

``` text
Subject: [PATCH] Revert incorrect default state for OrderListViewModel.ViewState
---
Index: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt
--- a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt	(revision 7b00c4367d1fd969a9fc51236d1aa0bef3f88e90)
+++ b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt	(date 1709759148606)
@@ -817,7 +817,7 @@
         val isSimplePaymentsAndOrderCreationFeedbackVisible: Boolean = false,
         val jitmEnabled: Boolean = false,
         val isErrorFetchingDataBannerVisible: Boolean = false,
-        val shouldDisplayTroubleshootingBanner: Boolean = false
+        val shouldDisplayTroubleshootingBanner: Boolean = true
     ) : Parcelable {
         @IgnoredOnParcel
         val isFilteringActive = filterCount > 0

```

## Success test scenario

1. Open the app, go to the Orders section
2. Once inside the Orders list, with the patch applied, the troubleshooting dialog should appear right away
3. Click the `troubleshooting` button
4. Verify a new screen opens with raw text data
5. Verify all checks suceeds

## Failure test scenario

1. Open the app, go to the Orders section
2. Once inside the Orders list, with the patch applied, the troubleshooting dialog should appear right away
3. Disable the Device internet connection
4. Click the `troubleshooting` button
5. Verify a new screen opens with raw text data
6. Verify all checks fails

Update release notes:
